File: system-test/dotnet/Dsl.Core/UseCase/Shop/UseCases/PublishCoupon.cs
using Common;
using Driver.Port.Shop;
using Driver.Port.Shop.Dtos;
using Driver.Port.Shop.Dtos.Error;
using Dsl.Core.Shared;
using Dsl.Core.Shop.UseCases.Base;

namespace Dsl.Core.Shop.UseCases;

public class PublishCoupon : BaseShopCommand<VoidValue, VoidVerification>
{
    private string? _code;
    private decimal? _discountRate;

    public PublishCoupon(IShopDriver driver, UseCaseContext context)
        : base(driver, context)
    {
    }

    public PublishCoupon Code(string? code)
    {
        _code = code;
        return this;
    }

    public PublishCoupon DiscountRate(decimal? discountRate)
    {
        _discountRate = discountRate;
        return this;
    }

    public override async Task<ShopUseCaseResult<VoidValue, VoidVerification>> Execute()
    {
        var request = new PublishCouponRequest
        {
            Code = _code,
            DiscountRate = _discountRate,
        };

        var result = await _driver.PublishCouponAsync(request);

        return new ShopUseCaseResult<VoidValue, VoidVerification>(
            result,
            _context,
            (response, ctx) => new VoidVerification(response, ctx));
    }
}

File: system-test/java/src/main/java/com/optivem/shop/dsl/core/usecase/shop/usecases/PublishCoupon.java
package com.optivem.shop.dsl.core.usecase.shop.usecases;

import com.optivem.shop.dsl.common.Converter;
import com.optivem.shop.dsl.core.shared.UseCaseContext;
import com.optivem.shop.dsl.core.shared.UseCaseResult;
import com.optivem.shop.dsl.core.shared.VoidVerification;
import com.optivem.shop.dsl.core.usecase.shop.usecases.base.BaseShopUseCase;
import com.optivem.shop.dsl.driver.port.shop.ShopDriver;
import com.optivem.shop.dsl.driver.port.shop.dtos.PublishCouponRequest;

public class PublishCoupon extends BaseShopUseCase<Void, VoidVerification> {
    private String code;
    private String discountRate;

    public PublishCoupon(ShopDriver driver, UseCaseContext context) {
        super(driver, context);
    }

    public PublishCoupon withCode(String code) {
        this.code = code;
        return this;
    }

    public PublishCoupon withDiscountRate(String discountRate) {
        this.discountRate = discountRate;
        return this;
    }

    public PublishCoupon withDiscountRate(double discountRate) {
        return withDiscountRate(Converter.fromDouble(discountRate));
    }

    @Override
    public UseCaseResult<Void, VoidVerification> execute() {
        var request = PublishCouponRequest.builder()
                .code(code)
                .discountRate(discountRate)
                .build();

        var result = driver.publishCoupon(request);
        return new UseCaseResult<>(result, context, VoidVerification::new);
    }
}
